llvm/lib/Target/DirectX/DXILResourceAccess.cpp

```diff
@@ -28,69 +28,88 @@
-    ++IndexIt;
-    Value *Offset = *IndexIt;
-    assert(++IndexIt == GEP->idx_end() && "Too many indices in GEP");
-    return Offset;
+    auto *GEP = dyn_cast<GetElementPtrInst>(Ptr);
+    assert(GEP && "Resource access through unexpected instruction");
+
+    unsigned NumIndices = GEP->getNumIndices();
+    uint64_t IndexScale = DL.getTypeAllocSize(GEP->getSourceElementType());
+    APInt ConstantOffset(DL.getIndexTypeSizeInBits(GEP->getType()), 0);
+    Value *GEPOffset;
+    if (GEP->accumulateConstantOffset(DL, ConstantOffset)) {
+      // We have a constant offset (in bytes).
+      GEPOffset =
+          ConstantInt::get(DL.getIndexType(GEP->getType()), ConstantOffset);
+      IndexScale = 1;
+    } else if (NumIndices == 1) {
+      // If we have a single index we're indexing into a top level array. This
+      // generally only happens with cbuffers.
+      GEPOffset = *GEP->idx_begin();
+    } else if (NumIndices == 2) {
+      // If we have two indices, this should be an access through a pointer.
+      auto IndexIt = GEP->idx_begin();
+      assert(cast<ConstantInt>(IndexIt)->getZExtValue() == 0 &&
+             "GEP is not indexing through pointer");
+      GEPOffset = *(++IndexIt);
+      assert(++IndexIt == GEP->idx_end() && "Too many indices in GEP");
+    } else
+      llvm_unreachable("Unhandled GEP structure for resource access");
+
+    uint64_t ElemSize = AccessSize;
+    if (!(IndexScale % ElemSize)) {
+      // If our scale is an exact multiple of the access size, adjust the
+      // scaling to avoid an unnecessary division.
+      IndexScale /= ElemSize;
+      ElemSize = 1;
+    }
+    if (IndexScale != 1)
+      GEPOffset = Builder.CreateMul(
+          GEPOffset, ConstantInt::get(Builder.getInt32Ty(), IndexScale));
+    if (ElemSize != 1)
+      GEPOffset = Builder.CreateUDiv(
+          GEPOffset, ConstantInt::get(Builder.getInt32Ty(), ElemSize));
+
+    Offset = Offset ? Builder.CreateAdd(Offset, GEPOffset) : GEPOffset;
+    Ptr = GEP->getPointerOperand();
   }
 
-  llvm_unreachable("Unhandled GEP structure for resource access");
+  llvm_unreachable("GEP of null pointer?");
 }
 
 static void createTypedBufferStore(IntrinsicInst *II, StoreInst *SI,
-                                   Value *Offset, dxil::ResourceTypeInfo &RTI) {
+                                   dxil::ResourceTypeInfo &RTI) {
+  const DataLayout &DL = SI->getDataLayout();
   IRBuilder<> Builder(SI);
   Type *ContainedType = RTI.getHandleTy()->getTypeParameter(0);
+  Type *ScalarType = ContainedType->getScalarType();
   Type *LoadType = StructType::get(ContainedType, Builder.getInt1Ty());
 
   Value *V = SI->getValueOperand();
   if (V->getType() == ContainedType) {
     // V is already the right type.
-    assert(!Offset && "store of whole element has offset?");
-  } else if (V->getType() == ContainedType->getScalarType()) {
+    assert(SI->getPointerOperand() == II &&
+           "store of whole element has offset?");
```

Comment:
I think this assert message should change?

Reply:
Nice catch. I've updated it.